### PR TITLE
Add `required_attributes` support for OAuth and auth assertion executor

### PR DESCRIPTION
**Pull request by thiva-k**
This pull request improves how required user attributes are determined and filtered in OAuth/OIDC authorization flows. It introduces logic to compute the minimal set of required attributes based on requested OIDC scopes and both ID token and access token configurations, ensuring only those attributes are included in authentication assertions. The changes are thoroughly tested with new unit and integration tests.

**Core logic and feature enhancements:**

* Added a new `getRequiredAttributes` function in `handler.go` that computes the required user attributes from OIDC scopes and application token config, combining ID token and access token attributes and filtering out non-required attributes.
* Updated the authorization flow to pass the computed `required_attributes` to the runtime data, ensuring downstream components use this filtered set.
* Modified the assertion generation logic to prefer `required_attributes` from runtime data, falling back to the application token config if not present.

**Testing improvements:**

* Added comprehensive unit tests for the new attribute computation logic, covering various scenarios and edge cases.
* Introduced an integration test to verify that only the required attributes (from scopes and token config) are present in the generated authentication assertion, and that non-required attributes are excluded.

**Supporting changes:**

* Added necessary imports (`strings`, `slices`, etc.) to support new logic in relevant files. [[1]](diffhunk://#diff-491fe16254a1b16f0310628b4961c9808c86dae266ea9797505cf7e4de29c83fR27) [[2]](diffhunk://#diff-de2c7e2e7e59fcb851b78ff55b0ed14c774e955787d303aa28ff80d2ddfccf23R26-R31) [[3]](diffhunk://#diff-ac3333d7ed31dfa0ac060343f3911e42cf2df252a9b1f14b1fccf9b621d9ce6cR27) [[4]](diffhunk://#diff-e8c794b8a25587e9999272480fa5d4e8c1fb7d8ac4198466c4836da5f646b9ffR22-R26)

These changes ensure that user attributes included in tokens are strictly limited to those required by the requested scopes and application configuration, improving both security and compliance.

### Related issue
- https://github.com/asgardeo/thunder/issues/1202

### CODE REVIEW

File: backend/internal/oauth/oauth2/authz/handler_test.go [hunk 1]
```diff
@@ -1288,3 +1289,173 @@ func (suite *AuthorizeHandlerTestSuite) TestHandleAuthorizePostRequest_Unsupport
...
+			name:       "Standard OIDC scopes with no IDToken config",
+			oidcScopes: []string{"openid", "profile", "email"},
+			app: &appmodel.OAuthAppConfigProcessedDTO{
+				Token: &appmodel.OAuthTokenConfig{
+					IDToken:     nil,
+					AccessToken: nil,
+				},
+			},
+			expectedResult: "sub name email email_verified",
+			description: "Should include all standard scope claims when IDToken.UserAttributes is empty",
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler_test.go]: Fixed

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 2]
```diff
@@ -544,3 +546,150 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+	// Expected: sub (from openid), name (from profile scope, filtered by IDToken.UserAttributes),
+	// email (from email scope, filtered by IDToken.UserAttributes), groups, roles (from AccessToken.UserAttributes)
+	// Should NOT contain: phone, customAttr, email_verified (not in IDToken.UserAttributes)
+
+	// Decode JWT to verify claims
+	claims := decodeJWTClaims(flowStep.Assertion, ts.T())
+
+	// Verify required attributes are present
+	ts.Assert().NotNil(claims["sub"], "sub claim should be present")
+	ts.Assert().Equal("requiredattrsuser", claims["sub"], "sub should match user ID")
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go]: Fixed

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 3]
```diff
@@ -621,3 +632,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+			requiredAttrsSet[attr] = true
+		}
+	}
+
+	// Convert set to slice
+	requiredAttrs := make([]string, 0, len(requiredAttrsSet))
+	for attr := range requiredAttrsSet {
+		requiredAttrs = append(requiredAttrs, attr)
+	}
+
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:707]: This is acceptable

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 4]
```diff
@@ -544,3 +546,153 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+
+// Helper function to extract auth ID from redirect location
+func extractAuthIDFromLocation(location string) string {
+	// Location format: /login?auth_id=xxx
+	u, err := url.Parse(location)
+	if err != nil {
+		return ""
+	}
+	return u.Query().Get("auth_id")
+}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go]: This is exsiting

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 5]
```diff
@@ -544,3 +545,143 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+		"name":                       "RequiredAttributesTestApp",
+		"client_id":                  "required_attrs_test_client",
+		"redirect_uris":              []string{scopeTestRedirectURI},
+		"grant_types":                []string{"authorization_code", "refresh_token"},
+		"response_types":             []string{"code"},
+		"token_endpoint_auth_method": "client_secret_basic",
+		"auth_flow_id":               ts.flowID,
+		"inbound_auth_config": []map[string]interface{}{
+			{
+				"type": "oauth",
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go]: Fixed

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 6]
```diff
@@ -544,3 +545,143 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+	authId, flowID, err := testutils.ExtractAuthData(location)
+	ts.Require().NoError(err, "Failed to extract auth data")
+	ts.Require().NotEmpty(authId, "Auth ID should not be empty")
+	ts.Require().NotEmpty(flowID, "Flow ID should not be empty")
+
+	// Step 2: Execute authentication flow
+	flowStep, err := testutils.ExecuteAuthenticationFlow(flowID, map[string]string{
+		"username": "requiredattrsuser",
+		"password": "TestPassword123!",
+	}, "")
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go]: Fixed

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 7]
```diff
@@ -621,3 +629,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+	return strings.Join(requiredAttrs, " ")
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:708]: This is acceptable

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 8]
```diff
@@ -621,3 +629,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+				scopeClaims = appClaims
+			}
+		}
+
+		// Fall back to standard OIDC scopes if no app-specific mapping
+		if scopeClaims == nil {
+			if standardScope, exists := oauth2const.StandardOIDCScopes[scope]; exists {
+				scopeClaims = standardScope.Claims
+			}
+		}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:685]: This will be improved in future effort

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 9]
```diff
@@ -544,3 +544,131 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+		"username": "requiredattrsuser",
+		"password": "TestPassword123!",
+		"email": "requiredattrs@test.com",
+		"firstName": "Required",
+		"lastName": "Attrs",
+		"name": "Required Attrs",
+		"groups": ["admin", "user"],
+		"roles": ["developer"],
+		"phone": "+1234567890",
+		"customAttr": "customValue"
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go:603]: Fixed

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 10]
```diff
@@ -248,37 +257,54 @@ func (a *authAssertExecutor) extractAuthenticatorReferences(
...
+			attr == oauth2const.ClaimOUName ||
+			attr == oauth2const.ClaimOUHandle {
+			continue
+		}
+
+		// Skip standard JWT claims if present in the user attributes
+		standardClaims := oauth2const.GetStandardClaims()
+		if slices.Contains(standardClaims, attr) {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go:277]: Fixed

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 11]
```diff
@@ -544,3 +545,150 @@ func (ts *OAuthAuthzScopeTestSuite) createTestAuthenticationFlow() string {
...
+		t.Fatalf("Failed to decode JWT payload: %v", err)
+	}
+
+	var claims map[string]interface{}
+	if err := json.Unmarshal(payload, &claims); err != nil {
+		t.Fatalf("Failed to unmarshal JWT claims: %v", err)
+	}
+
+	return claims
+}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply tests/integration/oauth/authz/authz_scope_test.go]: Fixed

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 12]
```diff
@@ -621,3 +629,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+
+		// Add claims to the set, but only if they're allowed in app config
+		for _, claim := range scopeClaims {
+			if len(idTokenAllowedAttrs) > 0 && slices.Contains(idTokenAllowedAttrs, claim) {
+				requiredAttrsSet[claim] = true
+			}
+		}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:692]: Fixed in auth assertion executor

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 13]
```diff
@@ -178,13 +181,18 @@ func (ah *authorizeHandler) handleInitialAuthorizationRequest(msg *OAuthMessage,
...
+		"requested_permissions": utils.StringifyStringArray(nonOidcScopes, " "),
+		"required_attributes":   requiredAttributes,
+	}
 	flowInitCtx := &flowexec.FlowInitContext{
 		ApplicationID: app.AppID,
 		FlowType:      string(flowcm.FlowTypeAuthentication),
-		RuntimeData: map[string]string{
-			"requested_permissions": utils.StringifyStringArray(nonOidcScopes, " "),
-		},
+		RuntimeData:   runtimeData,
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:194]: This is intentional

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 14]
```diff
@@ -157,9 +158,14 @@ func (a *authAssertExecutor) generateAuthAssertion(ctx *core.NodeContext, logger
...
+	// otherwise fallback to application token config
 	var userAttributes []string
-	if ctx.Application.Token != nil {
+	if requiredAttrs, exists := ctx.RuntimeData["required_attributes"]; exists && requiredAttrs != "" {
+		// Parse required_attributes
+		userAttributes = strings.Fields(requiredAttrs)
+	} else if ctx.Application.Token != nil {
+		// Fallback to application token user attributes
 		userAttributes = ctx.Application.Token.UserAttributes
 	}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go:169]: Updated

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 15]
```diff
@@ -631,3 +639,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+		}
+	}
+
+	// Convert set to slice
+	requiredAttrs := make([]string, 0, len(requiredAttrsSet))
+	for attr := range requiredAttrsSet {
+		requiredAttrs = append(requiredAttrs, attr)
+	}
+
+	return strings.Join(requiredAttrs, " ")
```
Comment (ThaminduDilshan) [inline backend/internal/oauth/oauth2/authz/handler.go:708]: I think this conversion is needed because runtime data can only be strings. Ideally runtime data can be anything (interface{}), and this conversion can be eliminated.

Shall we create an issue to track these improvements?

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 16]
```diff
@@ -631,3 +639,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+	if app == nil || app.Token == nil {
+		return ""
+	}
+
+	// Set to collect unique attribute names
+	requiredAttrsSet := make(map[string]bool)
+
+	// Map OIDC scopes to claims for ID token
+	var scopeClaimsMapping map[string][]string
+	var idTokenAllowedAttrs []string
```
Comment (ThaminduDilshan) [inline backend/internal/oauth/oauth2/authz/handler.go]: Shall we convert this also to be a slice (map of <string, bool>) so that lookup will be O(1) instead of O(n)?

```go
idTokenAllowedSet := make(map[string]bool, len(app.Token.IDToken.UserAttributes ))

for _, attr := range app.Token.IDToken.UserAttributes {
    idTokenAllowedSet[attr] = true
}

...

    for _, claim := range scopeClaims {
		if idTokenAllowedSet[claim] {
			requiredAttrsSet[claim] = true
		}
	}
```

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 17]
```diff
@@ -631,3 +639,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
+			if len(idTokenAllowedAttrs) > 0 && slices.Contains(idTokenAllowedAttrs, claim) {
+				requiredAttrsSet[claim] = true
+			}
+		}
+	}
+
+	// Add access token attributes from app config
+	if app.Token.AccessToken != nil &&
+		app.Token.AccessToken.UserAttributes != nil &&
+		len(app.Token.AccessToken.UserAttributes) > 0 {
```
Comment (ThaminduDilshan) [inline backend/internal/oauth/oauth2/authz/handler.go]:
```suggestion
		len(app.Token.AccessToken.UserAttributes) > 0 {
```

len() already checks for null values

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 18]
```diff
@@ -631,3 +639,63 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
 
 	return assertionClaims, authTime, nil
 }
+
+// getRequiredAttributes computes the required attributes based on OIDC scopes and access token config.
+func getRequiredAttributes(oidcScopes []string, app *appmodel.OAuthAppConfigProcessedDTO) string {
+	if app == nil || app.Token == nil {
+		return ""
+	}
```
Comment (ThaminduDilshan) [inline backend/internal/oauth/oauth2/authz/handler.go:646]: Shall we also check if access token and id token configs are defined? If neither is defined, there's no point of performing all other loops

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 19]
```diff
@@ -114,6 +115,9 @@ func (a *authAssertExecutor) generateAuthAssertion(ctx *core.NodeContext, logger
 	tokenSub := ""
 	if ctx.AuthenticatedUser.UserID != "" {
 		tokenSub = ctx.AuthenticatedUser.UserID
+	} else if sub, exists := ctx.RuntimeData[userAttributeSub]; exists && sub != "" {
+		// Fallback to runtime data for federated users
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/auth_assert_executor.go]: Is this logic correct according to the current implementation? Currently all federated users also still have an internal id which will be used as the authenticated user id. 
I feel like this is unnecessary

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 20]
```diff
@@ -157,9 +161,13 @@ func (a *authAssertExecutor) generateAuthAssertion(ctx *core.NodeContext, logger
...
+	// Get user attributes from required_attributes in runtimeData if present,
+	// otherwise fallback to application token config
 	var userAttributes []string
-	if ctx.Application.Assertion != nil {
+	if requiredAttrs, exists := ctx.RuntimeData["required_attributes"]; exists {
+		userAttributes = strings.Fields(requiredAttrs)
+	} else if ctx.Application.Assertion != nil {
+		// Fallback to application token user attributes
 		userAttributes = ctx.Application.Assertion.UserAttributes
 	}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go:169]: This is fine

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 21]
```diff
@@ -248,37 +256,55 @@ func (a *authAssertExecutor) extractAuthenticatorReferences(
...
+		// Check for the attribute in authenticated user attributes
+		if val, ok := ctx.AuthenticatedUser.Attributes[attr]; ok {
+			jwtClaims[attr] = val
+			continue
+		}
+
+		// For users with local accounts, fetch from user store
+		if ctx.AuthenticatedUser.UserID != "" && attrs == nil {
+			var err error
+			_, attrs, err = a.getUserAttributes(ctx.AuthenticatedUser.UserID)
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/auth_assert_executor.go:294]: IIUC, this logic fetches the userstore/ DB for each missing attribute. Shall we improve this logic to fetch the userstore only once?
Correct me if I'm wrong

Comment (ThaminduDilshan) [reply backend/internal/flow/executor/auth_assert_executor.go:294]:
```go
if ctx.AuthenticatedUser.UserID != "" && attrs == nil
```

Also is it correct/ isn't it redundant to have the `attrs == nil` check for the if condition?

Comment (ThaminduDilshan) [reply backend/internal/flow/executor/auth_assert_executor.go:294]: nvm. This logic is to fetch the userstore only once

File: backend/internal/flow/executor/auth_assert_executor.go [hunk 22]
```diff
@@ -248,37 +256,55 @@ func (a *authAssertExecutor) extractAuthenticatorReferences(
...
-				var err error
-				user, attrs, err = a.getUserAttributes(ctx.AuthenticatedUser.UserID)
-				if err != nil {
-					return err
-				}
-			}
-
-			// check for the attribute in user store attributes
-			if val, ok := attrs[attr]; ok {
+		// For federated users without local accounts, check runtime data as fallback
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/auth_assert_executor.go]: IMO this can be done for any user (both local and federated) and can be moved before making the userstore call to fetch user attributes. 
For example, if the mobile number is not found during sms otp authentication, it is fetched from the userstore and added to runtime data.

File: backend/internal/flow/executor/auth_assert_executor.go (hunk 19 above)
Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go]: If users are not provisioned before the auth_assert_executor, will the authenticated user id be available?

File: backend/internal/oauth/oauth2/authz/handler.go (hunk 18 above)
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:646]: Updated

File: backend/internal/flow/executor/auth_assert_executor.go (hunk 22 above)
Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go]: Updated

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 23]
```diff
@@ -178,13 +181,18 @@ func (ah *authorizeHandler) handleInitialAuthorizationRequest(msg *OAuthMessage,
 		oauthParams.RedirectURI = app.RedirectURIs[0]
 	}
 
+	// Compute required attributes from OIDC scopes and access token config
+	requiredAttributes := getRequiredAttributes(oidcScopes, app)
+
 	// Initiate flow with OAuth context
+	runtimeData := map[string]string{
+		"requested_permissions": utils.StringifyStringArray(nonOidcScopes, " "),
+		"required_attributes":   requiredAttributes,
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:189]: Fixed

File: backend/internal/oauth/oauth2/authz/handler.go (hunk 17 above)
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go]: Updated

File: backend/internal/oauth/oauth2/authz/handler.go (hunk 16 above)
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go]: Updated

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 24]
```diff
@@ -178,13 +180,18 @@ func (ah *authorizeHandler) handleInitialAuthorizationRequest(msg *OAuthMessage,
...
+	}
 	flowInitCtx := &flowexec.FlowInitContext{
 		ApplicationID: app.AppID,
 		FlowType:      string(flowcm.FlowTypeAuthentication),
-		RuntimeData: map[string]string{
-			"requested_permissions": utils.StringifyStringArray(nonOidcScopes, " "),
-		},
+		RuntimeData:   runtimeData,
 	}
```
[the remark this answers is not in the document]
Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go:195]: This is intentional

File: backend/internal/flow/executor/auth_assert_executor.go (hunk 19 above)
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/auth_assert_executor.go]: Current implementation doesn't allow authentication without a local user. Even if it's a just in provisioning, provisioning executor will be there before the auth assert executor, which will create an user and set the authenticated user id.

Comment (ThaminduDilshan) [reply backend/internal/flow/executor/auth_assert_executor.go]: And we can't hard code this to sub as well. There needs to be a way to configure subject attribute for the federated users (future improvement) and it should be handled through the related federated/ provisioning executor. Ideally that executor can set the authenticated user id based on the configured sub claim. 

So shall we remove this code?

Comment (thiva-k) [reply backend/internal/flow/executor/auth_assert_executor.go]: Removed

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 25]
```diff
@@ -631,3 +638,71 @@ func decodeAttributesFromAssertion(assertion string) (assertionClaims, time.Time
...
 }
+
+// getRequiredAttributes computes the required attributes based on OIDC scopes and access token config.
+func getRequiredAttributes(oidcScopes []string, app *appmodel.OAuthAppConfigProcessedDTO) string {
+	if app == nil || app.Token == nil {
+		return ""
+	}
+
+	// Early return if neither ID token nor access token configs are defined
+	if app.Token.IDToken == nil && app.Token.AccessToken == nil {
```
Comment (ThaminduDilshan) [inline backend/internal/oauth/oauth2/authz/handler.go]: Most of the times this check will be redundant right? All applications in our design will have id token/ access token configs. Rather shall we check whether the configured attributes are empty (len 0)?
Maybe I didn't mention it correctly in my previous comment

Comment (thiva-k) [reply backend/internal/oauth/oauth2/authz/handler.go]: Updated
